src/core/mac/mac_frame.cpp

```diff
@@ -145,6 +145,148 @@ ThreadError Frame::InitMacHeader(uint16_t aFcf, uint8_t aSecurityControl)
+        if ((fcf & Frame::kFcfPanidCompression) == 0)
+        {
+            VerifyOrExit((uint8_t)(GetPsduLength() - offset) > sizeof(PanId), ;);
+            offset += sizeof(PanId);
+        }
+
+        VerifyOrExit((uint8_t)(GetPsduLength() - offset) > sizeof(ShortAddress), ;);
+        offset += sizeof(ShortAddress);
+        break;
+
+    case Frame::kFcfSrcAddrExt:
+        if ((fcf & Frame::kFcfPanidCompression) == 0)
+        {
+            VerifyOrExit((uint8_t)(GetPsduLength() - offset) > sizeof(PanId), ;);
+            offset += sizeof(PanId);
+        }
+
+        VerifyOrExit((uint8_t)(GetPsduLength() - offset) > sizeof(ExtAddress), ;);
+        offset += sizeof(ExtAddress);
+        break;
+
+    default:
+        goto exit;
+    }
+
+    // Security Header
+    if (fcf & Frame::kFcfSecurityEnabled)
+    {
+        VerifyOrExit((uint8_t)(GetPsduLength() - offset) > sizeof(uint8_t), ;);
+
+        uint8_t secControl = GetPsdu()[offset];
+        offset += sizeof(uint8_t);
+
+        if (secControl & kSecLevelMask)
+        {
+            VerifyOrExit((uint8_t)(GetPsduLength() - offset) > kSecurityControlSize + kFrameCounterSize, ;);
+            offset += kSecurityControlSize + kFrameCounterSize;
+        }
+
+        switch (secControl & kKeyIdModeMask)
+        {
+        case kKeyIdMode0:
+            VerifyOrExit((uint8_t)(GetPsduLength() - offset) > kKeySourceSizeMode0, ;);
+            offset += kKeySourceSizeMode0;
+            break;
+
+        case kKeyIdMode1:
+            VerifyOrExit((uint8_t)(GetPsduLength() - offset) > kKeySourceSizeMode1 + kKeyIndexSize, ;);
+            offset += kKeySourceSizeMode1 + kKeyIndexSize;
+            break;
+
+        case kKeyIdMode2:
+            VerifyOrExit((uint8_t)(GetPsduLength() - offset) > kKeySourceSizeMode2 + kKeyIndexSize, ;);
+            offset += kKeySourceSizeMode2 + kKeyIndexSize;
+            break;
+
+        case kKeyIdMode3:
+            VerifyOrExit((uint8_t)(GetPsduLength() - offset) > kKeySourceSizeMode3 + kKeyIndexSize, ;);
+            offset += kKeySourceSizeMode3 + kKeyIndexSize;
+            break;
+        }
+
+        switch (secControl & kSecLevelMask)
+        {
+        case kSecNone:
+        case kSecEnc:
+            footerLength += kMic0Size;
+            break;
+
+        case kSecMic32:
+        case kSecEncMic32:
+            footerLength += kMic32Size;
+            break;
+
+        case kSecMic64:
+        case kSecEncMic64:
+            footerLength += kMic64Size;
+            break;
+
+        case kSecMic128:
+        case kSecEncMic128:
+            footerLength += kMic128Size;
+            break;
+        }
+    }
+
+    // Command ID
+    if ((fcf & kFcfFrameTypeMask) == kFcfFrameMacCmd)
+    {
+        VerifyOrExit((uint8_t)(GetPsduLength() - offset) > kCommandIdSize, ;);
+        offset += kCommandIdSize;
+    }
+
+    // Footer length
+    VerifyOrExit((uint8_t)(GetPsduLength() - offset) > footerLength, ;);
```

Comment:
I think these checks should be `>=` rather than `>`.  It's OK to have a MAC frame with zero payload.  #701 will depend on it.